### PR TITLE
add index_update method to backends

### PR DESCRIPTION
**Pull request by mganahl**
`index_update` allows to assign values to a tensor using a "mask":

```python 
x = backend.randn((10,))
mask = x>0.1
value = 0.0
backend.index_update(x, mask, value) #zero out all values > 0.1
```

### CODE REVIEW

**Comment by mganahl** [comment]
I took `index_update` from jax.If you don't like it I'm happy to rename.

**Comment by mganahl** [comment]
This PR already has the following PRs merged:
#375 -> #380 -> #382 
I'd recommend merging this in the above order to avoid clashes.

**Comment by Thenerdstation** [comment]
Why do you need this?

**Comment by mganahl** [comment]
There are cases where I need to take a pseudo inverse of a diagonal matrix (for stability reasons)
by zeroing those elements of the inverse for which the matrix elements are smaller than a cutoff.
